Temporary

bcoca opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
normalize 'temporary dir handling' 
this is partially done by also making shell plugins part of the new plugin config framework
fixes #31022

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
config
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```

bcoca committed:
allow shells to have per host options, remote_tmp

added language to shell
removed module lang setting from general as  plugins have it now
use get to avoid bad powershell plugin
more resilient tmp discovery, fall back to `pwd`
add shell to docs
fixed options for when frags are only options
added shell set ops in t_e and fixed option frags
normalize tmp dir usag4e

- pass tmpdir/tmp/temp options as env var to commands, making it default for tempfile
- adjusted ansiballz tmpdir
- default local tempfile usage to the configured local tmp
- set env temp in action

add options to powershell
shift temporary to internal envvar/params
ensure tempdir is set if we pass var
ensure basic and url use expected tempdir
ensure localhost uses local tmp
give /var/tmp priority, less perms issues
more consistent tempfile mgmt for ansiballz
made async_dir configurable
better action handling, allow for finally rm tmp
fixed tmp issue and no more tempdir in ballz
hostvarize world readable and admin users
always set shell tempdir
added comment to discourage use of exception/flow control

mattclay commented:
CI failure due to traceback:

```
Traceback (most recent call last):
  File "/root/ansible/bin/ansible-doc", line 58, in <module>
    import ansible.constants as C
  File "/root/ansible/lib/ansible/constants.py", line 189, in <module>
    os.environ['TMP'] = vars.get('DEFAULT_LOCAL_TMP')
AttributeError: 'builtin_function_or_method' object has no attribute 'get'
```

mattclay commented:
CI failure in most tests. One of the integration test failures due to: `Requested option remote_tmp was not defined in configuration`

mattclay commented:
CI failures in integration tests such as:

```
Traceback (most recent call last):
  File "/tmp/ansible-tmp-1512698752.05-21219899082320/command.py", line 112, in <module>
    temp_path = tempfile.mkdtemp(prefix='ansiballz_', dir=(os.environ.get('ANSIBLE_REMOTE_TEMP') or None))
  File "/usr/lib64/python2.7/tempfile.py", line 339, in mkdtemp
    _os.mkdir(file, 0700)
OSError: [Errno 13] Permission denied: '/tmp/ansible-tmp-1512698752.05-21219899082320/ansiballz_sm1xaG'
```

```
Traceback (most recent call last):
  File "<stdin>", line 112, in <module>
  File "/usr/lib/python2.7/tempfile.py", line 339, in mkdtemp
    _os.mkdir(file, 0700)
OSError: [Errno 2] No such file or directory: '~/.ansible/tmp/ansiballz_fjbPLg'
```

abadger commented on lib/ansible/constants.py:
```diff
@@ -183,3 +182,7 @@ def set_constant(name, value, export=vars()):
+
+    # speical hack to ensure all local temp files use the configured temp dir by default
+    if setting.name == 'DEFAULT_LOCAL_TMP':
+        os.environ['TEMPDIR'] = os.environ['TEMP'] = os.environ['TMP'] = value
```
This is not a good idea as it could break things that need a system tempdir instead of a per-user tempdir.

bcoca replied:
on controller side? i looked at all our tempfile uses and didn't see a case that would fit that scenario.
on remote we do have code that handles that case.

abadger replied:
Yeah, on controller-side.  for instance, if we have the local connection plugin with become to an unprivileged user, this might break any module that uses tempfile. (unless the become command always sets the environment variables to the value the suer we're becoming would expect.  But for sudo, at least, I think we let the user determine that via ansible.cfg settings).

abadger replied:
There's other places that could break too although those are more esoteric.  (We allow users to write their own plugins and changing TEMPDIR can affect the expectations of the tools that they might be invoking from the plugins).  It's the interaction of various different become commands with localhost that has me worried..

abadger replied:
Agreed today that we'll remove this and fix the callers to use 
`dir=C.DEFAULT_LOCAL_TMP`

Should be here:
```
 grep -ri tempfile . |grep import|grep -v ./modules|grep -v ./module_utils
./playbook/role/requirement.py:import tempfile
./executor/module_common.py:import tempfile
./executor/task_queue_manager.py:import tempfile
./plugins/action/__init__.py:import tempfile
./plugins/action/assemble.py:import tempfile
./plugins/action/copy.py:import tempfile
./plugins/action/template.py:import tempfile
./plugins/action/win_copy.py:import tempfile
./plugins/connection/winrm.py:import tempfile
./plugins/connection/paramiko_ssh.py:import tempfile
./plugins/connection/funcd.py:import tempfile
./galaxy/role.py:import tempfile
./parsing/dataloader.py:import tempfile
./parsing/vault/__init__.py:import tempfile
./config/manager.py:import tempfile
```

abadger commented on lib/ansible/executor/module_common.py:
```diff
@@ -304,11 +304,15 @@ def debug(command, zipped_mod, json_params):
-        # store this in remote_tmpdir (use system tempdir instead)
-        temp_path = tempfile.mkdtemp(prefix='ansible_')
+        # this temp path will be under the 'remote_tmp', in it's own subdir
+        tempdir = os.environ.get('ANSIBLE_REMOTE_TEMP', None) or None
```
Why do we need to do this instead of just substituting the temp dir into the python string?

abadger replied:
Hmm... Because you want to be able to set it differently per-host.... Okay.

bcoca replied:
can be done, but since i need the env var for other cases (async_wrapper) I stuck with that

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -814,11 +830,15 @@ def __init__(self, argument_spec, bypass_checks=False, no_log=False,
         self._options_context = list()
 
+        # set tempdir to remote tmp
+        self.tempdir = os.environ.get('ANSIBLE_REMOTE_TEMP', None)
```
We should do this via legal_inputs.  Yes, that will duplicate information that we had to pass to ansiballz via the environment but it keeps the handling of generic information from the controller to a single idiom.

abadger replied:
"[...] handling of generic information from the controller in AnsibleModule."

abadger replied:
Decided abadger will change this so that new-style modules get temporary from an _ansible parameter instead of from the environment.

abadger commented on lib/ansible/executor/module_common.py:
```diff
@@ -344,7 +348,10 @@ def debug(command, zipped_mod, json_params):
         try:
-            shutil.rmtree(temp_path)
+            if remove_tempdir:
+                shutil.rmtree(tempdir)
```
This might not be a good idea as it opens us up to a lot of race conditions.  I'd be tempted to just leave tempdir and only remove temp_path.  This means that when we are forced to makedir tempdir we need to give it relatively relaxed permissions.  (I think that's okay because temp_path has very strict permissions).

bcoca replied:
~~yes, it should be the local dir created and not the one by the controller, i'll 'rephrase'~~ nvmd, this was 'failsafe' because we previouslly removed the tempdir 1/2 the time, this code should not be needed at all at this point

mattclay commented:
CI failure in nearly all tests. Here's one traceback from ubuntu1604py3 integration tests:

```
Traceback (most recent call last):
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 524, in _execute
    result = self._handler.run(task_vars=variables)
  File "/root/ansible/lib/ansible/plugins/action/normal.py", line 45, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
  File "/root/ansible/lib/ansible/plugins/action/__init__.py", line 673, in _execute_module
    if not self._connection._shell.tempdir or 'tmp' not in tmp:
TypeError: argument of type 'NoneType' is not iterable
```

abadger commented:
Note: this is currently broken as the temporary directory needs to be writable by the becomed user. I think that fixupperms needs to be fixed to add write permissions.  When we do that, the allow_world_readable_tempfiles feature becomes untenable and will have to be replaced with something else (as we will need both world readable and world writable if we get to that fallback).

abadger commented:
Once that is fixed we'll have to test whether my becoming an unprivileged user with local connection concerns are really a problem or not.

abadger commented:
I think that wherever we can eliminate use of envvars we should.   Env vars
are really an external interface to a program.  Where possible we should
use something more targeted.

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -814,11 +830,15 @@ def __init__(self, argument_spec, bypass_checks=False, no_log=False,
+        self.tempdir = os.environ.get('ANSIBLE_REMOTE_TEMP', None)
+
+        # set for all tempfile calls
+        tempfile.tempdir = self.tempdir
```
This shouldn't be done.  Modules expect to get the system tempdir by calling tempfile's functions.  Modules are responsible for cleaning up after themselves.  If modules want to delegate that to us, they have the option of passing self.tempdir explicitly to the tempfile functions but we shouldn't be changing the default for them.  This can break other code (even library code) in ways that would be hard to debug.

abadger commented on lib/ansible/modules/utilities/logic/async_status.py:
```diff
@@ -57,8 +56,10 @@ def main():
+
     # setup logging directory
-    logdir = os.path.expanduser("~/.ansible_async")
+    logdir = os.path.expanduser(async_dir)
```
Do we need expandvars as well?  (and in async_wrapper)

abadger commented:
Making a list of things bcoca and I have talked about that need to be added as followup:
* Add ansible_SHELLPLUGIN_CONFIGNAME env vars so that you have the same level of granularity with env vars as config file.

abadger commented on lib/ansible/config/base.yml:
```diff
@@ -1,18 +1,6 @@
-  - {key: allow_world_readable_tmpfiles, section: defaults}
-  type: boolean
-  yaml: {key: defaults.allow_world_readable_tmpfiles}
-  version_added: "2.1"
```
This does not belong in shell.  Not unless we move fixup_perms2() into shell.

abadger commented on lib/ansible/errors/__init__.py:
```diff
@@ -252,11 +252,35 @@ def __init__(self, message="", obj=None, show_content=True, suppress_extended_er
                                                   suppress_extended_error=suppress_extended_error, orig_exc=orig_exc)
 
 
-class AnsibleActionSkip(AnsibleRuntimeError):
```
We decided that we don't like using exceptions as flow control.  But we don't need to fix this now.  We will add comments here that say that other people shouldn't copy this pattern.  We are going to change the action plugins that are using this to use a context manager to manager their temp directory as a replacement for this in the near future (2.6, probably) and revert the definition of these exceptions.

abadger replied:
this is done.

gundalow commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -287,11 +287,14 @@ def _make_tmp_path(self, remote_user=None):
+        self._connection._shell.tempdir = rc
+
+        if not use_system_tmp:
+            self._connection.shell.env.update('ANSIBLE_REMOTE_TEMP': self.tempdir)
```
Syntax error here which is is causing a lot of the failures

bcoca replied:
fixed, thnx

gundalow commented on lib/ansible/plugins/shell/__init__.py:
```diff
@@ -32,23 +31,33 @@
+
     def env_prefix(self, **kwargs):
-        env = self.env.copy()
         env.update(kwargs)
```
`env` no longer defined

jctanner commented:
ci_verified is waffling on this PR.

bot_skip

jctanner commented:
!waffling ci_verified

mattclay commented:
CI failure in integration tests due to traceback:

```
Traceback (most recent call last):
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 524, in _execute
    result = self._handler.run(task_vars=variables)
  File "/root/ansible/lib/ansible/plugins/action/assemble.py", line 149, in run
    remote_path = self._connection._shell.join_path(tmp, 'src')
  File "/root/ansible/lib/ansible/plugins/shell/__init__.py", line 64, in join_path
    return os.path.join(*args)
  File "/usr/lib/python3.5/posixpath.py", line 89, in join
    genericpath._check_arg_types('join', a, *p)
  File "/usr/lib/python3.5/genericpath.py", line 143, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'NoneType'
```

abadger commented:
@bcoca hey Brian, we're trying to get this merged.  If you're working on it too, jump onto slack or irc so that you can coordinate (and probably also tell us what we're doing wrong ;-)

abadger commented:
Hey @bcoca, since you're doing work on this but not communicating with the rest of the team, here's the work we've done so far: https://github.com/abadger/ansible/tree/temporary

We're hoping you get in touch with us soon so we can figure out how to get this merged before freeze.  Otherwise, on Monday we'll figure out whether to fork your branch at that current place or punt this to 2.7.

jborean93 commented on lib/ansible/plugins/shell/__init__.py:
```diff
@@ -141,20 +141,26 @@ def mkdtemp(self, basefile=None, system=False, mode=0o700, tmpdir=None):
+            # if present the user name is appended to resolve "that user's home"
+            user_home_path += username
+
         return 'echo "%s\t%spwd%s"' % (user_home_path, self._SHELL_SUB_LEFT, self._SHELL_SUB_RIGHT)
```
@bcoca this never worked for me, `echo "~username\tpwd` will return `~username /home/username` as the `~username` is inside double quotes.

mattclay commented:
CI failure due to multiple tracebacks:

```
Traceback (most recent call last):
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 524, in _execute
    result = self._handler.run(task_vars=variables)
  File "/root/ansible/lib/ansible/plugins/action/normal.py", line 45, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
  File "/root/ansible/lib/ansible/plugins/action/__init__.py", line 669, in _execute_module
    tmp = self._make_tmp_path()
  File "/root/ansible/lib/ansible/plugins/action/__init__.py", line 232, in _make_tmp_path
    admin_users = self._connection._shell.get_option('admin_users') + [remote_user]
  File "/root/ansible/lib/ansible/plugins/__init__.py", line 58, in get_option
    option_value = C.config.get_config_value(option, plugin_type=get_plugin_class(self), plugin_name=self._load_name, variables=hostvars)
  File "/root/ansible/lib/ansible/config/manager.py", line 284, in get_config_value
    value, _drop = self.get_config_value_and_origin(config, cfile=cfile, plugin_type=plugin_type, plugin_name=plugin_name, keys=keys, variables=variables)
  File "/root/ansible/lib/ansible/config/manager.py", line 304, in get_config_value_and_origin
    defs = self._plugins[plugin_type][plugin_name]
KeyError: 'shell'
```

```
Traceback (most recent call last):
  File "/private/var/root/ansible/lib/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/private/var/root/ansible/lib/ansible/executor/task_executor.py", line 524, in _execute
    result = self._handler.run(task_vars=variables)
  File "/private/var/root/ansible/lib/ansible/plugins/action/normal.py", line 45, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
  File "/private/var/root/ansible/lib/ansible/plugins/action/__init__.py", line 669, in _execute_module
    tmp = self._make_tmp_path()
  File "/private/var/root/ansible/lib/ansible/plugins/action/__init__.py", line 237, in _make_tmp_path
    tmpdir = self._remote_expand_user(self._connection._shell.get_option('remote_temp'), sudoable=False)
  File "/private/var/root/ansible/lib/ansible/plugins/action/__init__.py", line 547, in _remote_expand_user
    data = self._low_level_execute_command(self._connection._shell.expand_user(expand_path, username), sudoable=False)
  File "/private/var/root/ansible/lib/ansible/plugins/action/__init__.py", line 860, in _low_level_execute_command
    rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
  File "/private/var/root/ansible/lib/ansible/plugins/connection/local.py", line 95, in exec_command
    stderr=subprocess.PIPE,
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 711, in __init__
    errread, errwrite)
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1343, in _execute_child
    raise child_exception
TypeError: execv() arg 2 must contain only strings
```

I stopped looking after the first two.

abadger committed:
Mostly revert expand_user as it's not quite working.

This was an additional feature anyhow.

Kept the use of pwd as a fallback but moved it to a second ssh
connection.  This is not optimal but getting that to work in a single
ssh connection was part of the problem holding this up.

(cherry picked from commit 395b714120522f15e4c90a346f5e8e8d79213aca)

bcoca committed:
fixed script and other action plugins

ensure tmpdir deletion
allow for connections that don't support new options (legacy, 3rd party)
fixed tests

abadger commented:
rebuild_merge

abadger commented:
Merged for 2.5.  @jborean93 and I will get our other changes to this (as promised in the review orin the branch) submitted against the devel.

benjixx commented:
I'm getting the following traceback on devel, potentially due to this PR getting merged into devel:

```
Traceback (most recent call last):
  File "/home/travis/build/goodplay/goodplay/.tox/py27-ansibledevel/bin/ansible-playbook", line 118, in <module>
    exit_code = cli.run()
  File "/home/travis/build/goodplay/goodplay/.tox/py27-ansibledevel/lib/python2.7/site-packages/ansible/cli/playbook.py", line 122, in run
    results = pbex.run()
  File "/home/travis/build/goodplay/goodplay/.tox/py27-ansibledevel/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 89, in run
    self._tqm.load_callbacks()
  File "/home/travis/build/goodplay/goodplay/.tox/py27-ansibledevel/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 190, in load_callbacks
    for callback_plugin in callback_loader.all(class_only=True):
  File "/home/travis/build/goodplay/goodplay/.tox/py27-ansibledevel/lib/python2.7/site-packages/ansible/plugins/loader.py", line 462, in all
    self._update_object(obj, name, path)
  File "/home/travis/build/goodplay/goodplay/.tox/py27-ansibledevel/lib/python2.7/site-packages/ansible/plugins/loader.py", line 346, in _update_object
    self._load_config_defs(name, path)
  File "/home/travis/build/goodplay/goodplay/.tox/py27-ansibledevel/lib/python2.7/site-packages/ansible/plugins/loader.py", line 217, in _load_config_defs
    if 'options' in dstring and isinstance(dstring['options'], dict):
TypeError: argument of type 'NoneType' is not iterable
```

abadger commented:
@benjixx Please open a bug for that and link to this PR since you think it's probably a change in here somewhere.  @bcoca is on vacation and I don't have time to look into the problem today but definitely don't want it to get forgotten.

sivel commented:
@benjixx your traceback does not align with the current code in devel.  The code in your traceback was changed in https://github.com/ansible/ansible/commit/869a3184922 which landed in September, far before this PR.
